StashRepository: Clean up comment sorting, add comments

proski opened the pull request:
When extracting parameters from comments, start with the oldest comments,
as the code iterates over all comments. Parameters from the new comments
are processed last, overwriting older values.

When deleting old "BUILD FINISHED" comments, don't sort the comments, it
doesn't matter in which order they are deleted.

When looking for comments to decide whether to run a build, start with
the most recent comments. The loop terminates early as soon as it's clear
whether to start the build or not. Don't sort and reverse the sort order
as two operations, use reverse sorting.

proski commented:
This is a very simple change, mostly intended to replace commented out code with proper comments.

But looking at the code made me see several inefficiencies:

* It should not be necessary to read all comments at once. A smarter iterator would read the next portion of comments only after exhausting the current batch, reducing the load on the Bitbucket Server, especially for PRs that don't need to be built.
* Reading comments to decide whether to build the PR and extracting parameters from the comments can be combined. If the code decides to build the PR, processing the comments should continue until all parameters defined for the project are populated or until the earliest comment.
* Regular expressions should be precompiled. They should not be compiled inside loops.

jakub-bochenski commented:
Looks good, but again: I'm a bit afraid to change it without some unit test coverage (just a simple case with 2 comments to check they are still processed in proper order)

jakub-bochenski reviewed (changes requested):
see above

proski committed:
StashRepository: Clean up comment sorting, add comments

When extracting parameters from comments, start with the oldest comments,
as the code iterates over all comments. Parameters from the new comments
are processed last, overwriting older values.

When deleting old "BUILD FINISHED" comments, don't sort the comments, it
doesn't matter in which order they are deleted.

When looking for comments to decide whether to run a build, start with
the most recent comments. The loop terminates early as soon as it's clear
whether to start the build or not. Don't sort and reverse the sort order
as two operations, use reverse sorting.

proski commented:
Added unit tests for all affected methods.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -214,4 +244,70 @@ public void getTargetPullRequestsBuildsIfSkipPhraseIsNull() {
+  }
+
+  @Test
+  public void getAdditionalParametersUsesNewestParameterDefinitionForNewCommentsFirst() {
```
I think the names might be a little confusing here. I would name the previous case `getAdditionalParametersUsesNewestParameterDefinition` and this one `getAdditionalParametersUsesNewestParameterDefinitionRegardlessOfListOrder`

proski replied:
Done.
